### DISCUSSION TITLE
Example of pyproject.toml for a namespace package?

### DISCUSSION THREAD

**Discussion by waydegilliam** [pdm-project/pdm#600] [Q&A] [answered]
Hello!

I'm trying to *editable* install a local package using a namespace, however I'm doing something wrong in my pyproject.toml (either for the namespace package I'm trying to install or the non-namespace package I'm installing the namespace package to) as the namespace isn't getting recognized but only the package name. For example, "company_name.core_package" can only be imported as "core_package". 

The pyproject.toml for the local namespace package:
```
[project]
name = "company_name.core"
version = "0.0.1"
description = ""
authors = [{name = "my name", email = "myname@gmail.com"}]
readme = "README.md"
requires-python = ">=3.8"
dependencies = []

[project.urls]
homepage = ""

[tool.pdm]
package-dir = "company_name"
includes = ["core/*"]
```

The pyproject.toml for the package I'm trying to *editable* install for. FYI this is just a package that install jupyter notebook dependencies along with other packages and doesn't have any modules:
```
[project]
name = "jupyter_scratchpad"
version = "0.0.1"
description = ""
authors = [{name = "my name", email = "myname@gmail.com"}]
readme = "README.md"
requires-python = ">=3.8"
dependencies = [
    "notebook~=6.4",
    "-e file:///absolute/path/to/core#egg=company_name.core",
]

[project.urls]
homepage = ""
```

I haven't found any examples in the official docs or on GitHub, so any help would be appreciated!

**Comment by frostming** [comment, marked as answer]
The file structure of the namespace package should be:

```
company_name
    core
        __init__.py
        ...
pyproject.toml
```
The `includes` value of the projects should be `includes = ["company_name"]`

The namespace package is `company_name` instead of `core`, i.e. you can have `company_name.utils`, `company_name.others` as standalone packages, all belong to the namespace `company_name`. The project name `company_name.core` just acts as an identifier, with no special meanings. You can name it as `company_name_core` and it will still be imported with `import company_name.core`. The file structure is what changes the rule.

See https://github.com/pdm-project/pdm-pep517 as an example

**Comment by criccomini** [comment]
Adding a note for those that find this: It should be `[tool.pdm.build]` not `[tool.pdm]`. See [here](https://daobook.github.io/pdm/pyproject/tool-pdm/#implicit-namespace-packages) for more info.
